File: SWEA/D3/5215. 햄버거 다이어트/햄버거 다이어트.py
def dfs(idx, score, stk):
    global answer
    answer = max(answer, score)

    for i in range(idx, N):
        if visited[i]:
            continue
        if stk + arr[i][1] > L:
            continue
        visited[i] = True
        dfs(i, score + arr[i][0], stk + arr[i][1])
        visited[i] = False


global answer
T = int(input())
for test_case in range(1, T + 1):
    answer = 0
    N, L = map(int, input().split())
    visited = [False] * N
    arr = [list(map(int, input().split())) for _ in range(N)]
    dfs(0, 0, 0)

    print(f"#{test_case} {answer}")

File: SWEA/D3/5215. 햄버거 다이어트/햄버거 다이어트.java
import java.io.*;
import java.util.*;

class Solution {
    static int N, L;
    static int[][] ingredients;
    static int answer;

	public static void main(String args[]) throws Exception
	{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;
        
        int T = Integer.parseInt(br.readLine());
		for(int test_case = 1; test_case <= T; test_case++) {
            st = new StringTokenizer(br.readLine());
            N = Integer.parseInt(st.nextToken());
            L = Integer.parseInt(st.nextToken());

            ingredients = new int[2][N];
            answer = 0;

            for (int i = 0; i < N; i++) {
                st = new StringTokenizer(br.readLine());
                ingredients[0][i] = Integer.parseInt(st.nextToken());
                ingredients[1][i] = Integer.parseInt(st.nextToken());
            }

            dfs(0, 0, 0);
            
			sb.append(String.format("#%d %d\n", test_case, answer));
		}
        
        bw.write(sb.toString());
        bw.flush();
        bw.close();
        br.close();
	}

    static void dfs(int idx, int score, int calorie) {
        if (idx == N) {
            answer = Math.max(answer, score);
            return;
        }

        int newCalorie = calorie + ingredients[1][idx];
        if (newCalorie <= L) {
            dfs(idx + 1, score + ingredients[0][idx], newCalorie);
        }
        
        dfs(idx + 1, score, calorie);
    }
}